src/sync/sync_permission.cpp

```diff
@@ -0,0 +1,295 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync_permission.hpp"
+#include "sync_user.hpp"
+#include "sync_config.hpp"
+#include "sync_manager.hpp"
+#include "sync_session.hpp"
+#include "object_schema.hpp"
+#include "impl/object_accessor_impl.hpp"
+
+#if REALM_PLATFORM_APPLE
+#include <realm/util/cf_ptr.hpp>
+#else
+#include <sole.hpp>
+#endif
+
+#include "property.hpp"
+
+using namespace realm;
+
+namespace {
+
+// TODO: this should already exist somewhere.
+struct MallocBuffer {
+    void* buffer_ptr;
+
+    MallocBuffer(size_t bytes)
+    : buffer_ptr(malloc(bytes))
+    {
+        if (!buffer_ptr)
+            throw std::runtime_error("Could not successfully malloc memory.");
+    }
+
+    ~MallocBuffer()
+    {
+        free(buffer_ptr);
+    }
+
+    template<typename T>
+    T* get()
+    {
+        return static_cast<T*>(buffer_ptr);
+    }
+};
+
+std::string make_uuid()
+{
+#if REALM_PLATFORM_APPLE
+    CFPtr<CFUUIDRef> uuid = adoptCF(CFUUIDCreate(NULL));
+    CFPtr<CFStringRef> uuid_str = adoptCF(CFUUIDCreateString(NULL, uuid.get()));
+    size_t buffer_len = CFStringGetLength(uuid_str.get()) + 1;
+    MallocBuffer buffer(buffer_len);
+    CFStringGetCString(uuid_str.get(), buffer.get<char>(), buffer_len, kCFStringEncodingASCII);
+    std::string uuid_cpp_str = buffer.get<char>();
+    return uuid_cpp_str;
+#else
+    return sole::uuid4().str();
+#endif
+}
+
+}
+
+size_t PermissionResults::size() {
+    REALM_ASSERT(m_results->size());
+    return m_results->size() - 1;
+}
+
+#pragma mark - Permission
+
+Permission::Permission(const Permission& p)
```

Comment:
Yeah, I think once there are implementations for `Permission::Condition`'s copy constructor/assignment operator, the equivalents for `Permission` should be auto generated.